File: ArithmeticSlices.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Oct 18 10:18:38 2019

@author: tanvirkaur
"""

# time complexity = O(n)
# space complexity = O(n)
# the sequence must contain atleat three elements therefore we start iterating from the third element till the last element and store the corresponding values in the tmp array.
# it is dp because we are using the prev values to update the new values.
class Solution:
    def numberOfArithmeticSlices(self, A: List[int]) -> int:
        tmp = [0]*len(A)
        for i in range(2,len(A)):
            if A[i]-A[i-1] == A[i-1]-A[i-2]:
                tmp[i] = tmp[i-1]+1
        return sum(tmp)

File: ArithmeticSlices.java
package DynamicProgramming8;

public class ArithmeticSlices {
    /* Created by palak on 8/16/2021 */

    /**
     Top Down space O(1)
     Time Complexity: O(n)
     */
    public int numberOfArithmeticSlices(int[] nums) {
        if(nums == null || nums.length < 0) return 0;
        int n = nums.length;
        int sum = 0;
        // int [] dp = new int[n];

        int curr = 0;
        int prev = 0;

        for(int i = 2 ; i < n ; i++) {
            if(nums[i] - nums[i - 1] == nums[i - 1] - nums[i - 2]) {
                curr = prev + 1;
                sum += curr;
                prev = curr;
            } else {
                prev = 0;
            }
        }

        return sum;
    }

    public static void main(String[] args) {

    }
}
